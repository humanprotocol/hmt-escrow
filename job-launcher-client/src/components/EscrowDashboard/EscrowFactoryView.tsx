import * as React from 'react';
import { Box, Typography } from '@mui/material';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import { CardTextBlock } from '../Cards';
import Events from './Events';

interface IEscrow {
  title: string;
  address: string;
  eventsUrl: string;
  scanner: string;
  scannerUrl?: string;
  latestEscrow?: string;
  count?: number;
  pendingEventCount?: number;
  bulkTransferEventCount?: number;
  intermediateStorageEventCount?: number;
  totalEventCount?: number;
}

export const EscrowFactoryView: React.FC<IEscrow> = ({
  title,
  count,
  address,
  latestEscrow,
  eventsUrl,
  scanner,
  pendingEventCount,
  bulkTransferEventCount,
  intermediateStorageEventCount,
  totalEventCount,
}): React.ReactElement => {
  return (
    <Box>
      <Typography>{title}</Typography>
      <Card variant="outlined">
        <CardContent>
          <CardTextBlock title="Address" value={address} />
          <CardTextBlock title="Latest Escrow" value={latestEscrow} />
          <CardTextBlock title="Amount of jobs" value={count} />
          <CardTextBlock
            title="All Escrows Pending Events"
            value={pendingEventCount}
          />
          <CardTextBlock
            title="All Escrows BulkTransfer Events"
            value={bulkTransferEventCount}
          />
          <CardTextBlock
            title="All Escrows IntermediateStorage Events"
            value={intermediateStorageEventCount}
          />
          <CardTextBlock
            title="Total Number Of Escrows Events"
            value={totalEventCount}
          />
          <Events url={eventsUrl} scanner={scanner} />
        </CardContent>
      </Card>
    </Box>
  );
};
