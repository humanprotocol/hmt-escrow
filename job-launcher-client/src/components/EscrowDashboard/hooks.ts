import * as React from 'react';
import { toast } from 'react-toastify';

import { ethers } from 'ethers';
import { ESCROW_STATS_RAW, fetchRawQuery } from '../../queries';
import { formatHexToNumber } from './helpers';
import { INetwork } from '../../constants/networkConstants';

const EscrowFactoryABI = require('../../contracts/EscrowFactoryABI.json');

export const useEscrowHook = ({
  scanner,
  scannerUrl,
  rpcUrl,
  defaultFactoryAddr: address,
  graphqlClientUrl,
}: INetwork) => {
  const eventsUrl = `${scanner}/address/${address}#events`;

  const [latestEscrow, setLatestEscrow] = React.useState<string>();
  const [count, setCount] = React.useState<number>();
  const [pendingEventCount, setPendingEventCount] = React.useState<number>();
  const [bulkTransferEventCount, setBulkTransferEventCount] =
    React.useState<number>();
  const [intermediateStorageEventCount, setIntermediateStorageEventCount] =
    React.useState<number>();
  const [totalEventCount, setTotalEventCount] = React.useState<number>();

  React.useEffect(() => {
    async function setupEscrow() {
      try {
        const iface = new ethers.utils.Interface(EscrowFactoryABI);
        const httpProvider = new ethers.providers.JsonRpcProvider(rpcUrl);
        const Contract = new ethers.Contract(address, iface, httpProvider);

        const lastEscrow = await Contract.lastEscrow();
        setLatestEscrow(lastEscrow);

        const { _hex } = await Contract.counter();
        const counter = formatHexToNumber({ hex: _hex, radix: 16 });
        setCount(counter);

        try {
          const { data: { escrowStatistics: stats = {} } = {} } =
            await fetchRawQuery(graphqlClientUrl, ESCROW_STATS_RAW);

          setPendingEventCount(stats.pendingEventCount);
          setBulkTransferEventCount(stats.bulkTransferEventCount);
          setIntermediateStorageEventCount(stats.intermediateStorageEventCount);
          setTotalEventCount(
            Number(stats.pendingEventCount) +
              Number(stats.bulkTransferEventCount) +
              Number(stats.intermediateStorageEventCount)
          );
        } catch (err) {
          setPendingEventCount(Number.NaN);
          setBulkTransferEventCount(Number.NaN);
          setIntermediateStorageEventCount(Number.NaN);
          setTotalEventCount(Number.NaN);
        }
      } catch (err) {
        const error: any = err;
        toast.error(error?.message, {
          position: 'top-right',
        });
        console.log(error);
      }
    }
    setupEscrow();
  }, [address, rpcUrl, graphqlClientUrl]);

  return {
    address,
    eventsUrl,
    scanner,
    scannerUrl,
    latestEscrow,
    count,
    pendingEventCount,
    bulkTransferEventCount,
    intermediateStorageEventCount,
    totalEventCount,
  };
};
