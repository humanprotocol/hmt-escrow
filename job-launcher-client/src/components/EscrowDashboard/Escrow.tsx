import * as React from 'react';
import { INetwork } from '../../constants/networkConstants';

import { EscrowFactoryView } from './EscrowFactoryView';
import { useEscrowHook } from './hooks';

interface IEscrowContainer {
  network: INetwork;
}

export const Escrow: React.FC<IEscrowContainer> = ({
  network,
}): React.ReactElement => {
  const escrowData = useEscrowHook(network);

  return <EscrowFactoryView title={network.title} {...escrowData} />;
};
