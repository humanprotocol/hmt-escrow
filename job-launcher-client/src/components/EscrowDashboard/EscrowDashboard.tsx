import { Box } from '@mui/material';
import * as React from 'react';
import { networks } from '../../constants/networkConstants';
import { Escrow } from './Escrow';

export const EscrowDashboard: React.FC = (): React.ReactElement => {
  return (
    <Box
      sx={{
        display: 'flex',
        justifyContent: 'space-evenly',
        flexWrap: 'wrap',
      }}
    >
      {networks
        .filter((network) => !network.isDeprecated)
        .map((network) => (
          <Box mx={1} mb={2} key={network.key}>
            <Escrow network={network} />
          </Box>
        ))}
    </Box>
  );
};
