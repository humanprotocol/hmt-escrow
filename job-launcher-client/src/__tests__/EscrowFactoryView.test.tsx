import React from "react";
import { render, screen } from "@testing-library/react";
import { act, create } from "react-test-renderer";
import { EscrowFactoryView } from "../components/EscrowDashboard/EscrowFactoryView";

const mock = {
  title: "Polygon Mainnet",
  address: "0x45eBc3eAE6DA485097054ae10BA1A0f8e8c7f794",
  latestEscrow: "eaac4d45c3c41f449cf7c94622afacbc",
  eventsUrl: "test- url",
  scanner: "test-scanner",
  count: 10,
  pendingEventCount: 10,
  bulkTransferEventCount: 10,
  intermediateStorageEventCount: 10,
  totalEventCount: 30,
};

describe("when rendered with a `title` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
      />
    );
    expect(screen.getByText(mock.title)).toBeInTheDocument();
  });
});

describe("when rendered with a `count` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        count={mock.count}
      />
    );
    expect(screen.getByText(mock.count)).toBeInTheDocument();
  });
});

describe("when rendered with a `address` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
      />
    );
    expect(screen.findByLabelText(mock.address)).toBeTruthy();
  });
});

describe("when rendered with a `latestEscrow` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        latestEscrow={mock.latestEscrow}
      />
    );
    expect(screen.findByLabelText(mock.latestEscrow)).toBeTruthy();
  });
});

describe("when rendered with a `eventsUrl` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
      />
    );
    expect(screen.findByLabelText(mock.eventsUrl)).toBeTruthy();
  });
});

describe("when rendered with a `scanner` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
      />
    );
    expect(screen.findByLabelText(mock.scanner)).toBeTruthy();
  });
});

describe("when rendered with a `pendingEventCount` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        pendingEventCount={mock.pendingEventCount}
      />
    );
    expect(screen.findByLabelText(mock.pendingEventCount)).toBeTruthy();
  });
});

describe("when rendered with a `intermediateStorageEventCount` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        intermediateStorageEventCount={mock.intermediateStorageEventCount}
      />
    );
    expect(
      screen.findByLabelText(mock.intermediateStorageEventCount)
    ).toBeTruthy();
  });
});

describe("when rendered with a `bulkTransferEventCount` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        bulkTransferEventCount={mock.bulkTransferEventCount}
      />
    );
    expect(screen.findByLabelText(mock.bulkTransferEventCount)).toBeTruthy();
  });
});

describe("when rendered with a `totalEventCount` prop", () => {
  it("should paste it into the component", () => {
    render(
      <EscrowFactoryView
        title={mock.title}
        address={mock.address}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
        totalEventCount={mock.totalEventCount}
      />
    );
    expect(screen.findByLabelText(mock.totalEventCount)).toBeTruthy();
  });
});

it("renders correctly", () => {
  let root;
  act(() => {
    root = create(
      <EscrowFactoryView
        title={mock.title}
        count={mock.count}
        address={mock.address}
        latestEscrow={mock.latestEscrow}
        eventsUrl={mock.eventsUrl}
        scanner={mock.scanner}
      />
    );
  });
  expect(root).toMatchSnapshot();
});
