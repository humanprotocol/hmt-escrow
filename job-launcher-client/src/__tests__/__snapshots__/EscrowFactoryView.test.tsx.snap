// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiBox-root css-0"
>
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    Polygon Mainnet
  </p>
  <div
    className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
  >
    <div
      className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Address
          </p>
        </span>
      </div>
      <p
        className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
      >
        0x45eBc3eAE6DA485097054ae10BA1A0f8e8c7f794
      </p>
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Latest Escrow
          </p>
        </span>
      </div>
      <p
        className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
      >
        eaac4d45c3c41f449cf7c94622afacbc
      </p>
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Amount of jobs
          </p>
        </span>
      </div>
      <p
        className="MuiTypography-root MuiTypography-body2 css-1ypfif-MuiTypography-root"
      >
        10
      </p>
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            All Escrows Pending Events
          </p>
        </span>
      </div>
      <div
        className="MuiBox-root css-1l4w6pd"
      >
        <span
          className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style={
            {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={{}}
            />
          </svg>
        </span>
      </div>
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            All Escrows BulkTransfer Events
          </p>
        </span>
      </div>
      <div
        className="MuiBox-root css-1l4w6pd"
      >
        <span
          className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style={
            {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={{}}
            />
          </svg>
        </span>
      </div>
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            All Escrows IntermediateStorage Events
          </p>
        </span>
      </div>
      <div
        className="MuiBox-root css-1l4w6pd"
      >
        <span
          className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style={
            {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={{}}
            />
          </svg>
        </span>
      </div>
      <div
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
        role="separator"
      >
        <span
          className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
        >
          <p
            className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Total Number Of Escrows Events
          </p>
        </span>
      </div>
      <div
        className="MuiBox-root css-1l4w6pd"
      >
        <span
          className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style={
            {
              "height": "1.5rem",
              "width": "1.5rem",
            }
          }
        >
          <svg
            className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx={44}
              cy={44}
              fill="none"
              r={20.2}
              strokeWidth={3.6}
              style={{}}
            />
          </svg>
        </span>
      </div>
      <div
        className="MuiBox-root css-xi606m"
      >
        <div
          className="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-1ws8119-MuiDivider-root"
          role="separator"
        >
          <span
            className="MuiDivider-wrapper css-qywfm8-MuiDivider-wrapper"
          >
            <p
              className="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
            >
              All deployed escrows
            </p>
          </span>
        </div>
        <a
          className="MuiTypography-root MuiTypography-inherit MuiTypography-alignCenter MuiLink-root MuiLink-underlineAlways css-8wztv5-MuiTypography-root-MuiLink-root"
          href="test- url"
          onBlur={[Function]}
          onFocus={[Function]}
          rel="noopener noreferrer"
          target="_blank"
        >
          test-scanner
        </a>
        <p
          className="MuiTypography-root MuiTypography-body2 css-1ncpc4g-MuiTypography-root"
        >
          Each event has a payload of ERC20 token address and Escrow Address
        </p>
        <p
          className="MuiTypography-root MuiTypography-body2 css-1ncpc4g-MuiTypography-root"
        >
          Change the type of the second argument to "Address" to see an Escrow address
        </p>
      </div>
    </div>
  </div>
</div>
`;
